File: kernel/src/cpio.zig
// Functions and data structures to decode the cpio format (used in initrd)

const std = @import("std");

pub const CpioNode = extern struct {
    magic: [6]u8 = magic_marker,
    device: [6]u8,
    i_number: [6]u8,
    mode: [6]u8,
    user_id: [6]u8,
    group_id: [6]u8,
    num_links: [6]u8,
    r_device: [6]u8,
    modified_time: [11]u8,
    name_len: [6]u8,
    file_len: [11]u8,

    const name_offset = 76;
    const magic_const: [6]u8 = [_]u8{ '0', '7', '0', '7', '0', '7' };
};

pub fn octalToBinary(string: []const u8) u64 {
    var number: u64 = 0;
    for (string) |digit| {
        number <<= 3;
        number += digit -% '0';
    }
    return number;
}

test "Octal To Binary" {
    std.testing.expectEqual(0o070707, comptime octalToBinary("070707"));
    std.testing.expectEqual(0o32, comptime octalToBinary("000032"));
    std.testing.expectEqual(0o4040, comptime octalToBinary("00000004040"));
}

pub fn cpioFindFile(archive: []const u8, file_name: []const u8) ?[]const u8 {
    var cur_pos: usize = 0;
    loop: while (cur_pos < archive.len) {
        const node = @ptrCast(*const CpioNode, &archive[cur_pos]);
        const node_file_size = octalToBinary(node.file_len[0..node.file_len.len]);
        const node_name_len = octalToBinary(node.name_len[0..node.name_len.len]);
        const node_file_name = archive[cur_pos + CpioNode.name_offset .. cur_pos + CpioNode.name_offset + node_name_len - 1];
        // Check magic
        {
            comptime var i: usize = 0;
            inline while (i < node.magic.len) : (i += 1) {
                if (node.magic[i] != CpioNode.magic_const[i]) {
                    return null;
                }
            }
        }
        // Check file name
        {
            if (node_name_len - 1 != file_name.len) {
                cur_pos += 76 + node_name_len + node_file_size;
                continue :loop;
            }
            var i: usize = 0;
            while (i < node_name_len - 1) : (i += 1) {
                if (node_file_name[i] != file_name[i]) {
                    cur_pos += CpioNode.name_offset + node_name_len + node_file_size;
                    continue :loop;
                }
            }
        }
        return archive[cur_pos + CpioNode.name_offset + node_name_len .. cur_pos + CpioNode.name_offset + node_name_len + node_file_size];
    }
    return null;
}

File: kernel_rust/src/cpio.rs
use core::mem::size_of;

#[repr(C)]
#[derive(Clone, Copy)]
pub struct Node {
    pub magic: [u8; 6],
    pub device: [u8; 6],
    pub i_number: [u8; 6],
    pub mode: [u8; 6],
    pub user_id: [u8; 6],
    pub group_id: [u8; 6],
    pub num_links: [u8; 6],
    pub r_device: [u8; 6],
    pub modified_time: [u8; 11],
    name_len_octal: [u8; 6],
    file_size_octal: [u8; 11],
}

impl Node {
    pub const NAME_OFFSET: usize = 76;
    pub const MAGIC: &[u8; 6] = b"070707";

    /// Returns the length of the node's ASCII name plus the NULL byte at the end.
    pub fn get_name_cstring_len(&self) -> usize {
        octal_to_binary(&self.name_len_octal)
    }

    pub fn get_file_size(&self) -> usize {
        octal_to_binary(&self.file_size_octal)
    }
}

pub fn octal_to_binary(octal: &[u8]) -> usize {
    let mut number = 0;
    for digit in octal {
        number <<= 3;
        number += (digit - b'0') as usize;
    }
    number
}

pub fn find_file<'a>(archive: &'a [u8], file_name: &[u8]) -> Option<&'a [u8]> {
    let mut current_pos = 0;
    while current_pos < archive.len() {
        let node = unsafe {
            // Check enough bytes exist for a Node
            if archive.len() - current_pos + 1 < size_of::<Node>() {
                break;
            }
            (&archive[current_pos] as *const u8 as *const Node)
                .as_ref()
                .unwrap_unchecked()
        };
        let node_name_len = node.get_name_cstring_len();
        let node_file_size = node.get_file_size();
        let node_name = &archive[current_pos + Node::NAME_OFFSET..][0..node_name_len - 1];
        // Check magic
        if &node.magic != Node::MAGIC {
            break;
        }
        // Check file name
        if file_name != node_name {
            current_pos += 76 + node_name_len + node_file_size;
            continue;
        }
        return Some(
            &archive[current_pos + Node::NAME_OFFSET + node_name_len..][0..node_file_size],
        );
    }
    None
}
